### ISSUE TITLE
Slow bundle download(~2 mins) in dev mode by ReloadCommand(r) on iOS after RN 72 upgrade

### ISSUE THREAD

**Issue by jayshah123** [react/metro#1243]
### Description

I have a large bundle.
After recent upgrade to `RN 0.72.12`, metro version - `metro@0.76.9` (Also tried with RN `0.73.9` and Metro `v0.80.8`), I see large download times for already traversed bundle, when using reload command "r"

The first load, bundle does not need downloading. 

But for every reload command("r" press on iOS),
the bundle download to simulator (for the already traversed bundle) takes about 2 mins.

I did some analysis on size of the bundle transferred by Metro Server is ~248 MB. I am working to reduce bundle size, but due to lack of tree-shaking, the current size is about ~248 MB. In the absensece of tree shaking, what tools are recommended for bundle size analysis?

The size was measured by the value of following code in `Server.js` in metro source code.
```js
mres.setHeader(
          "Content-Length",
          String(Buffer.byteLength(result.bundle))
        );
```

For the same project, if I try to download using curl, it takes about 7s to download to a file, for an already traversed bundle, which was measured via:
```
curl -kv -w '\n* Response time: %{time_total}s\n' http://localhost:8081/index.bundle?platform=ios&dev=true&minify=false&inlineSourceMap=false&modulesOnly=false&runModule=true&app=com.mypackagename
```

### Steps to reproduce

1. install the application which has a really large bundle
2. let the bundle load completely.
3. Press "r" to reload.
4. You will see Downloading "1..100%" takes around 2 mins to finish.

### React Native Version

0.72.12

### Affected Platforms

Runtime - iOS

### Output of `npx react-native info`

```text
System:
  OS: macOS 14.3.1
  CPU: (10) arm64 Apple M1 Max
  Memory: 7.24 GB / 32.00 GB
  Shell:
    version: "5.9"
    path: /bin/zsh
Binaries:
  Node:
    version: 18.13.0
    path: ~/.nvm/versions/node/v18.13.0/bin/node
  Yarn:
    version: 1.18.0
    path: ~/.nvm/versions/node/v18.13.0/bin/yarn
  npm:
    version: 8.19.3
    path: ~/.nvm/versions/node/v18.13.0/bin/npm
  Watchman:
    version: 2024.01.22.00
    path: /opt/homebrew/bin/watchman
Managers:
  CocoaPods:
    version: 1.14.3
    path: /Users/jshah/.rbenv/shims/pod
SDKs:
  iOS SDK:
    Platforms:
      - DriverKit 23.2
      - iOS 17.2
      - macOS 14.2
      - tvOS 17.2
      - visionOS 1.0
      - watchOS 10.2
  Android SDK: Not Found
IDEs:
  Android Studio: 2022.3 AI-223.8836.35.2231.10811636
  Xcode:
    version: 15.2/15C500b
    path: /usr/bin/xcodebuild
Languages:
  Java:
    version: 17.0.10
    path: /usr/bin/javac
  Ruby:
    version: 3.3.0
    path: /Users/jshah/.rbenv/shims/ruby
npmPackages:
  "@react-native-community/cli": Not Found
  react:
    installed: 18.2.0
    wanted: 18.2.0
  react-native:
    installed: 0.72.12
    wanted: 0.72.12
  react-native-macos: Not Found
npmGlobalPackages:
  "*react-native*": Not Found
Android:
  hermesEnabled: true
  newArchEnabled: false
iOS:
  hermesEnabled: true
  newArchEnabled: false
```

### Stacktrace or Logs

```text
NA
```

### Reproducer

private code/bundle.

### Screenshots and Videos

Downloading 1..100% takes ~2 mins when ReloadCommand "r" is pressed.

<img width="443" alt="Screenshot 2024-03-23 at 10 15 53 AM" src="https://github.com/facebook/react-native/assets/659202/3f5dd65b-1c09-4dd4-9967-5774dc7d5f73">

**Comment by efstathiosntonas** [comment]
@jayshah123 you could try [react-native-bundle-visualizer f](https://github.com/IjzerenHein/react-native-bundle-visualizer)

**Comment by jayshah123** [comment]
Bundle size does not seem to be a problem, due to following observation:

**Good case -  kill the app by swipe from recents, relaunch from apps** - Downloading bundle takes ~5 secs.

On the other hand,
**Bad case - "r key press"** - Downloading bundle takes ~2 mins.

Maintainers, what is a good place/way to start investigating the root cause.

**Comment by 509dave16** [comment]
@jayshah123 I'm having the same problem on my local network when connecting to Metro from Pixel 4 or iPhone 15 plus through an Xfinity gateway. The IP address of my 2023 MacBook Pro is a little different, 10.0.0.113. But that shouldn't matter. Tried turning off Private Relay on both Apple devices. Tried internet sharing between Apple devices.

It's 54 MB bundle that normally takes 72 seconds to load. Turning off my work VPN on the MacBook helped get it down to 30 seconds. But that's still way too long :( .

**Comment by zhangshaoju1987** [comment]
I face the same problem.
http://192.168.1.121:8081/index.bundle?platform=ios&dev=true&lazy=true&minify=false&inlineSourceMap=false&modulesOnly=false&runModule=true&app=space.zhumi.app
I type the url in my pc chrome, and it keeped a loading continuous text output status. it takes half an hour.
![image](https://github.com/user-attachments/assets/77a09962-bc86-48a0-8ba9-4112d22ae27c)

**Comment by montaserfzy** [comment]
> Now on RN version 0.73.9, same issue persists.

Same here

**Comment by robhogan** [comment]
Could someone make a screen recording of the output of Metro during this slow request, starting Metro with `DEBUG=Metro:* yarn start` to show debug output?

Bonus; the debug output should show the bundle URL being requested by the app. Repeat the test using `curl` to get download the bundle and compare the speed. If it’s fast with `curl`, it’s an RN rather than Metro issue.

**Comment by jayshah123** [comment]
As an additional information, I am attaching both metro config and recording for metro loading + curl fetch.

Here is my Metro config:
```js
/**
 * Metro configuration for React Native
 * https://github.com/facebook/react-native
 *
 * @format
 */
const { getDefaultConfig } = require('@expo/metro-config');

const path = require('path');

async function getConfig(appDir) {
  const config = await getDefaultConfig(__dirname);
  // setup watch folders
  config.watchFolders = [path.resolve(appDir, 'node_modules')];
  // setup transformer
  config.transformer = {
    ...config.transformer,
    babelTransformerPath: require.resolve('react-native-svg-transformer/expo'),
    getTransformOptions: async () => ({
      transform: {
        experimentalImportSupport: true,
        inlineRequires: true,
      },
    }),
  };

  // setup resolver source extensions
  const originalSourceExts = config.resolver.sourceExts;
  let newSourceExts = process.env.RN_SRC_EXT
    ? process.env.RN_SRC_EXT.split(',').concat(originalSourceExts)
    : originalSourceExts;
  config.resolver.sourceExts = [...newSourceExts, 'svg', 'cjs', 'mjs'];

  // setup resolver asset extensions
  const origianlAssetExts = config.resolver.assetExts;
  config.resolver.assetExts = [
    ...origianlAssetExts.filter((ext) => ext !== 'svg'),
    'css',
    'scss',
  ];

  // setup resolver extra node modules
  config.resolver.extraNodeModules = {
    stream: require.resolve('readable-stream'),
  };

  return config;
}

module.exports = getConfig(__dirname);

```

And here is the video for metro loading:
Notice following:
1. First load time : bundling + downloading is slow
2. Shake > Reload - Downloading time slow on iOS.
3. In the end I curl the same bundles - really fast.

https://github.com/user-attachments/assets/dfc1ca07-7a60-4017-ad71-0660d72cc818

**Comment by jayshah123** [comment]
Let me know if any additional information is needed.

**Comment by jpodpro** [comment]
Incredibly slow for me too. about 1% every 30 seconds. Completely unusable.

**Comment by brentforder** [comment]
For me, this issue is a nightmare for testing iOS builds on MacOS.
Environment: RN 0.76.5, Metro 0.81.0, Expo 52.0.0.21 on MacOS Sequoia 15.2. MacOS Firewall is disabled (all network activity is permitted).
Android testing on Windows 11 is fine... Bundle downloads take a few seconds at most.

Metro's bundling task itself completes in 10s or less, but the file-streaming / file-serving of the bundle from MacOS after the bundle is created takes an insane amount of time.
In my case, the "downloading" with a hardware iPhone takes around an hour. I have to painstakingly plan each testing build to be highly likely to succeed, otherwise I'm not making any progress with this platform. Test-driven development is  completely out of the question.

I tried using Metro's tunnel option to work around this, but it looks like that is also broken. Metro silently fails with the tunnel option, while the ngrok tunnel is never actually created (the generated ngrok URL returns an error indicating that the tunnel doesn't exist).

**Comment by brentforder** [comment]
I've found a workaround, at least for my case!
When I use my iPhone as a WiFi hotspot and disable my other network options in MacOS Settings, downloading the bundle from Metro is now super fast as expected.

**Comment by jpodpro** [comment]
@brentforder Does this require the device to have a SIM data connection?

**Comment by brentforder** [comment]
> @brentforder Does this require the device to have a SIM data connection?

Well that makes me wonder if there is some way to use a WiFi network as a connection source for a hotspot, but I've never tried that. In my case, I'm using a device with a SIM data connection.

**Comment by Sylchauf** [comment]
Same issue here … 

I have found a workaround. Install Tailscale on both (VPN), then try to fire the menu in the app on real device (shake or m button if it's connected to metro)
Then, in «Configure bundler» type the IP of MacOS through the VPN (100.x.x.x)

Fast again !

**Comment by 509dave16** [comment]
This method has worked off and on for me @Sylchauf, @brentforder, @jpodpro , @jayshah123, @montaserfzy, @zhangshaoju1987:

If the `index.js` bundle from the Metro bundler server is taking more than 5 seconds to download to your physical iOS device, consider using [macOS internet sharing](https://support.apple.com/guide/mac-help/share-internet-connection-mac-network-users-mchlp1540/mac). Some engineers have reported going from 60+ second download times to 3 second downloads:

- **Connect** your iPhone to your computer via USB/USB-C
  - Very important for later step in configuring Internet Sharing
- **Disable** WiFi on your iPhone
  - Internet will be shared from computer's WiFi over USB/USB-C.
- **Open** "Internet Sharing" via Spotlight Search or the Settings app
- **Click** the info icon next to "Internet Sharing"
  - ![Image](https://github.com/user-attachments/assets/3a424103-62b7-4f10-ba26-382616010275)

- **Make sure** you:
  - Select `Wi-Fi` for "Share your connection from"
  - Enable `iPhone USB` for "To devices using"
  - Toggle "Internet Sharing" `On` and confirm this on next prompt by clicking `Turn On`
     - ![Image](https://github.com/user-attachments/assets/aae436b7-69df-4d78-b21a-e7cdef1b9529)
- **Done**: Now try reopening the local build of the app on your iPhone to see if it loads from Metro faster
- **NOTE**: Technically you should not have to configure "Internet Sharing" again on your computer. But if it forgets your configuration, just follow the steps again.

**Comment by de-tester** [comment]
For me problem was that the phone got into the different VLAN on the same router. After the iOS upgrade device was dropped from trusted network to the default VLAN, which resulted notebook was still in trusted vlan 192.168.3.* and iPhone was in the default 192.168.1.* which caused that effect. As soon as both devices got in the same subnet everything works as before.

TLDR: Make sure that both devices are on the same subnet even if the network is the same.

**Comment by robhogan** [comment]
Thanks for sharing tips folks. The local `curl` above, plus reports that it's working fine on Android, proves there was no issue with the Metro server. If people are still experiencing this, the RN repo is a better place to report.
